torchtune/modules/common_utils.py

```diff
@@ -48,3 +54,118 @@ def reparametrize_as_dtype_state_dict_post_hook(
             state_dict[k] = v.to(dtype)
             if offload_to_cpu:
                 state_dict[k] = state_dict[k].cpu()
+
+
+# mmap.MAP_SHARED is not supported on Windows but this change targets colab.
+if torch.__version__ >= "2.5.0.dev20240906" and not sys.platform == "win32":
+
+    def _low_ram_reparametrize_as_dtype_state_dict_post_hook(
+        model: nn.Module,
+        state_dict: Dict[str, Any],
+        *args: Tuple[Any, ...],
+        dtype: torch.dtype = torch.bfloat16,
+        offload_to_cpu: bool = True,
+        **kwargs: Dict[Any, Any],
+    ):
+        """
+        A state_dict hook that replaces NF4 tensors with their restored
+        higher-precision weight and optionally offloads the restored weight to CPU.
+        Use this hook to avoid increased peak GPU memory usage during checkpoint
+        save when training with QLoRA.
+
+        This hook is similar to ``reparametrize_as_dtype_state_dict_post_hook`` but uses
+        FakeTensor and mmap(2) to avoid CPU OOM on colab.
+
+        This function is meant to be used with PyTorch's ``nn.Module._register_state_dict_hook``, i.e.
+
+        >>> m = MyModule()
+        >>> m._register_state_dict_hook(reparametrize_as_dtype_state_dict_post_hook)
+
+        If the hook is registered per the above process, this hook will be called _after_ the module's
+        ``state_dict`` method is called. The hook will replace all ``NF4Tensor`` instances by unquantizing
+        them to the original dtype, and optionally offload the restored weight to CPU.
+
+        Args:
+            model (nn.Module): the model to take ``state_dict()`` on
+            state_dict (Dict[str, Any]): the state dict to modify
+            *args (Tuple[Any, ...]): Unused args passed when running this as a state_dict hook.
+            dtype (torch.dtype): the dtype to restore the weight to. Default is ``torch.bfloat16``.
+            offload_to_cpu (bool): whether to offload the restored weight to CPU. Default is ``True``.
+            **kwargs (Dict[Any, Any]): Unused keyword args passed when running this as a state_dict hook.
+        """
+        # Create a state dict of FakeTensors that matches the state_dict
+        mode = FakeTensorMode()
+        converter = FakeTensorConverter()
+        fake_state_dict = OrderedDict()
+        for k, v in state_dict.items():
+            if isinstance(v, NF4Tensor):
+                fake_state_dict[k] = converter.from_real_tensor(mode, v).to(dtype)
+            else:
+                fake_state_dict[k] = converter.from_real_tensor(mode, v)
+
+            if offload_to_cpu:
+                fake_state_dict[k] = fake_state_dict[k].cpu()
+
+        # Create a state_dict on disk with space reserved for storage bytes
+        # Then load with mmap and MAP_SHARED (can writeback to disk file)
+        dest_state_dict_path = "/tmp/fake_state_dict.pt"
+        with torch.serialization.skip_data(materialize_fake_tensors=True):
```

Comment:
noob q: what does `materialize_fake_tensors` mean in this context?

Reply:
It means that FakeTensors in the object passed to `torch.save` will be treated as if they were real tensors. The implications are

- `torch.load` will load a tensor on the FakeTensor's device with storage allocated but uninitialized (0s)